File: HW/3_Quera/Q3/3_reshte_parantezi.py
class Stack:
    def __init__(self):
        self.list = []

    def push(self, element):
        self.list.append(element)

    def pop(self):
        return self.list.pop()

    def is_empty(self):
        if self.list.__len__() == 0:
            return True
        else:
            return False


stack_a = Stack()

temp = 0
q = int(input())

parentheses = list(str(input()))


beautifully = []

for i in parentheses:
    if i == '(':
        if stack_a.is_empty():
            temp = 0
        else:
            temp += 1
        stack_a.push(i)
    else:
        temp += 1
        stack_a.pop()
        # if stack_a.is_empty():
        #     temp += 1
        beautifully.append(temp)

print(max(beautifully))

File: HW/3_Quera/Q3/3_reshte_parantezi.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Scanner;
import java.util.Stack;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = Integer.parseInt(scanner.nextLine());
        String input = scanner.nextLine();
        Stack<Character> stack = new Stack<Character>();
        int temp = 0;
        ArrayList<Integer> maxes = new ArrayList<>();
        for(Character c : input.toCharArray()){
            if(c == '(')
                if (stack.isEmpty()) {
                    maxes.add(temp);
                    temp = 0;
                    stack.push('(');
                }
                else {
                    temp ++;
                    stack.push('(');
                }
            else {
                stack.pop();
                temp ++;
            }
        }

        System.out.print(Collections.max(maxes));
    }
}
